service/efsPort.py

```diff
@@ -0,0 +1,630 @@
+from eos.db import gamedata_session, getItemsByCategory, getCategory, getAttributeInfo, getGroup
+from eos.gamedata import Category, Group, Item, Traits, Attribute, Effect, ItemEffect
+from logbook import Logger
+pyfalog = Logger(__name__)
+eos.db.saveddata_meta.create_all()
+
+
+class RigSize(Enum):
+    # Matches to item attribute "rigSize" on ship and rig items
+    SMALL = 1
+    MEDIUM = 2
+    LARGE = 3
+    CAPITAL = 4
+
+
+class EfsPort():
+    wepTestSet = {}
+
+    @staticmethod
+    def attrDirectMap(values, target, source):
+        for val in values:
+            target[val] = source.itemModifiedAttributes[val]
+
+    @staticmethod
+    def getT2MwdSpeed(fit, sFit):
+        fitID = fit.ID
+        propID = None
+        shipHasMedSlots = fit.ship.itemModifiedAttributes["medSlots"] > 0
+        shipPower = fit.ship.itemModifiedAttributes["powerOutput"]
+        # Monitors have a 99% reduction to prop mod power requirements
+        if fit.ship.name == "Monitor":
+            shipPower *= 100
+        rigSize = fit.ship.itemModifiedAttributes["rigSize"]
+        if not shipHasMedSlots:
+            return None
+
+        filterVal = Item.groupID == getGroup("Propulsion Module").ID
+        propMods = gamedata_session.query(Item).options().filter(filterVal).all()
+        mapPropData = lambda propName: \
+                      next(map(lambda propMod: {"id": propMod.typeID, "powerReq": propMod.attributes["power"].value},
+                               (filter(lambda mod: mod.name == propName, propMods))))
+        mwd5mn = mapPropData("5MN Microwarpdrive II")
+        mwd50mn = mapPropData("50MN Microwarpdrive II")
+        mwd500mn = mapPropData("500MN Microwarpdrive II")
+        mwd50000mn = mapPropData("50000MN Microwarpdrive II")
+        if rigSize == RigSize.SMALL or rigSize is None:
+            propID = mwd5mn["id"] if shipPower > mwd5mn["powerReq"] else None
+        elif rigSize == RigSize.MEDIUM:
+            propID = mwd50mn["id"] if shipPower > mwd50mn["powerReq"] else mwd5mn["id"]
+        elif rigSize == RigSize.LARGE:
+            propID = mwd500mn["id"] if shipPower > mwd500mn["powerReq"] else mwd50mn["id"]
+        elif rigSize == RigSize.CAPITAL:
+            propID = mwd50000mn["id"] if shipPower > mwd50000mn["powerReq"] else mwd500mn["id"]
+
+        if propID is None:
+            return None
+        sFit.appendModule(fitID, propID)
+        sFit.recalc(fit)
+        fit = eos.db.getFit(fitID)
+        mwdPropSpeed = fit.maxSpeed
+        mwdPosition = list(filter(lambda mod: mod.item and mod.item.ID == propID, fit.modules))[0].position
+        sFit.removeModule(fitID, mwdPosition)
+        sFit.recalc(fit)
+        fit = eos.db.getFit(fitID)
+        return mwdPropSpeed
+
+    @staticmethod
+    def getPropData(fit, sFit):
+        fitID = fit.ID
+        propGroupId = getGroup("Propulsion Module").ID
+        propMods = filter(lambda mod: mod.item and mod.item.groupID == propGroupId, fit.modules)
```

Comment:
Could just do `mod.item.group.name == 'Propulsion Module'` instead of having to get the group first, but doesn't really matter. Six of one and half a dozen of the other.

Additionally... I'm seeing a lot of `filter/lambda` combinations for narrowing lists down. The more conventional / pythonic way would be to use list comprehension. So, instead of

`propMods = filter(lambda mod: mod.item and mod.item.groupID == propGroupId, fit.modules)`

it would be something like

`propMods = [mod for mod in fit.modules if mod.item and mod.item.groupID == propGroupId]`

Reply:
Yeah I was looking at the db itself and wasn't aware that was a thing for most of my time making this. I'll change it accordingly.

I'll look into swapping some of the lambdas out.